### PR TITLE
Fix working with read buffers in StreamingFormatExecutor

### PR DESCRIPTION
**Pull request by Avogar**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix working with read buffers in StreamingFormatExecutor, previously it could lead to segfaults in Kafka and other streaming engines.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit da76d78cfb16675d064fe41634bd4faa97672551 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57438/da76d78cfb16675d064fe41634bd4faa97672551/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/7066239140">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57438/da76d78cfb16675d064fe41634bd4faa97672551/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Processors/Formats/IInputFormat.cpp [hunk 1]
```diff
@@ -24,7 +24,6 @@ void IInputFormat::resetParser()
 
 void IInputFormat::setReadBuffer(ReadBuffer & in_)
 {
-    chassert(in); // not supported by random-access formats
```
Comment (novikd) [inline src/Processors/Formats/IInputFormat.cpp:27]: why?

File: src/Interpreters/AsynchronousInsertQueue.cpp [hunk 2]
```diff
@@ -783,11 +782,6 @@ Chunk AsynchronousInsertQueue::processEntriesWithParsing(
         auto buffer = std::make_unique<ReadBufferFromString>(*bytes);
         size_t num_bytes = bytes->size();
         size_t num_rows = executor.execute(*buffer);
-
-        /// Keep buffer, because it still can be used
-        /// in destructor, while resetting buffer at next iteration.
-        last_buffer = std::move(buffer);
```
Comment (novikd) [inline src/Interpreters/AsynchronousInsertQueue.cpp:789]: That's actually a great change!

**Review by novikd — COMMENTED** [review]
In general LGTM

File: src/Processors/Formats/IInputFormat.cpp (hunk 1 above)
Comment (Avogar) [reply src/Processors/Formats/IInputFormat.cpp:27]: Because before `setReadBuffer` now we call `resetReadBuffer` that resets `in` to `nullptr`

**Comment by Avogar** [comment]
Test `test_storage_kafka/test.py::test_issue26643` is failed in CI, but locally it works...
I will try to investigate

**Comment by Avogar** [comment]
Finally I found out why kafka tests are failing in the CI

**Review by novikd — APPROVED** [review]

**Pull request merged**